### DISCUSSION TITLE
How do i achieve dots on the line graph for series entry?

### DISCUSSION THREAD

**Discussion by vivekgupta4Git** [patrykandpatrick/vico#609] [Questions] [answered]
<img width="867" alt="Screenshot 2024-03-11 at 7 39 42 PM" src="https://github.com/patrykandpatrick/vico/assets/91813403/78746a24-d353-4ed2-8844-7d5f63597306">

So my requirement is dots for each series entry on the line graph and background for threshold min and max.Till now, I am abled to change the line spec to straight line by providing my own implementation of PointConnector. And also, taking an example from the sample of threshold (copy paste the rememberThreshold() method and its requirement) I was able to use the decorations for the minimum and maximum threshold background (sample had drawn background in front of the series but my requirement was behind the series so changed few line of codes in the Threshold data class) . This is what i achieved right now 
<img width="497" alt="Screenshot 2024-03-11 at 7 57 22 PM" src="https://github.com/patrykandpatrick/vico/assets/91813403/0355a680-11f1-4a44-b351-11420d6d2662"> So how do i achieve dots and threshold min and max line (dashed line)  with background?
Any help or direction will be appreciated!.

**Comment by Gowsky** [comment]
Hello @vivekgupta4Git, since you have a custom implementation of `ThresholdLine`, you can start by adding a `LineComponent` that uses a `DashedShape` like the following.
````kotlin
val dashedLine = LineComponent(
    color = Color.GREEN,
    shape = DashedShape(dashLengthDp = 4f, gapLengthDp = 4f),
    thicknessDp = 1f,
)
````
Then somewhere at the end of the drawing function e.g. after `lineComponent.draw(...)` add the following code.
````kotlin
dashedLine.drawHorizontal(
    context = context,    
    left = bounds.left,   
    right = bounds.right, 
    centerY = topY,       
)
dashedLine.drawHorizontal(
    context = context,    
    left = bounds.left,   
    right = bounds.right, 
    centerY = bottomY,       
)
````
Alternatively you can add two other `ThresholdLine`s (for the top and bottom values) that would use the `dashedLine` as their `lineComponent`s.

**Reply by vivekgupta4Git** [reply]
@Gowsky Thanks for the input. I have achieved the desired output including circle on the series but It felt like a hack though :P I created three circle differ by 1f in radius after drawing Line in the implementation of PointConnecter. Btw, What is the preferred way of creating threshold lines?

**Reply by patrickmichalik** [reply, marked as answer]
There’s a dedicated API for line points, @vivekgupta4Git. Below is a quick example with Vico 2 and Jetpack Compose. Let me know if you need an example for Vico 1 or for the view system.

```kt
val modelProducer = remember { CartesianChartModelProducer.build() }
LaunchedEffect(Unit) { modelProducer.tryRunTransaction { lineSeries { series(2f, 4f, 3f, 6f) } } }
CartesianChartHost(
    rememberCartesianChart(
        rememberLineCartesianLayer(
            listOf(
                rememberLineSpec(
                    shader = remember { DynamicShaders.color(Color.Blue) },
                    point = rememberShapeComponent(Shapes.pillShape, Color.Blue),
                    pointSize = 8.dp,
                ),
            ),
        ),
        startAxis = rememberStartAxis(),
        bottomAxis = rememberBottomAxis(),
    ),
    modelProducer,
)
```

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/fa6d1d4d-552a-4821-8bd8-b8dee2685184" width="720" />

Regarding `ThresholdLine`s, could you expand on the question? Would you like to know which of the two solutions mentioned by @Gowsky is preferable?

**Reply by vivekgupta4Git** [reply]
I am asking for a dedicated api to create a threshold area/line as I couldn't find rememberThresholdLine() in the library but in the sample. Is there any?

**Reply by vivekgupta4Git** [reply]
@patrickmichalik Thanks for the example, thats exactly what I wanted. :)

**Reply by patrickmichalik** [reply]
Regarding `ThresholdLine` creation, @vivekgupta4Git, the API for this is the class constructor. The sample app’s `rememberThresholdLine` function uses the constructor and customizes the `ThresholdLine` for a specific sample chart. A composable factory function for `ThresholdLine` would be expected in the library API, and we’ve addressed this in [Vico 2.0.0 Alpha 12](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.12). We’ve split `ThresholdLine` into two new `Decoration`s, `HorizontalBox` and `HorizontalLine`, and added `rememberHorizontalBox` and `rememberHorizontalLine` functions to the `compose` module.

As for the example, I’m glad it helped. You’re welcome!
